src/vs/workbench/parts/debug/common/debugModel.ts

```diff
@@ -90,16 +91,61 @@ export function getFullExpressionName(expression: debug.IExpression, sessionType
 }
 
 export class Thread implements debug.IThread {
-
+	private promisedCallStack: TPromise<debug.IStackFrame[]>;
+	private cachedCallStack: debug.IStackFrame[];
+	private callStackQueried: boolean;
 	public stoppedDetails: debug.IRawStoppedDetails;
+	public stopped: boolean;
 
-	constructor(public name: string, public threadId, public callStack: debug.IStackFrame[]) {
+	constructor(public name: string, public threadId) {
+		this.promisedCallStack = undefined;
 		this.stoppedDetails = undefined;
+		this.callStackQueried = false;
+		this.cachedCallStack = undefined;
+		this.stopped = false;
 	}
 
 	public getId(): string {
 		return `thread:${ this.name }:${ this.threadId }`;
 	}
+
+	public clearCallStack(): void {
+		this.promisedCallStack = undefined;
+		this.cachedCallStack = undefined;
+		this.callStackQueried = false;
+	}
+
+	public getCachedCallStack(): debug.IStackFrame[] {
+		return this.cachedCallStack;
+	}
+
+	public getCallStack(debugService: debug.IDebugService): TPromise<debug.IStackFrame[]> {
+		if (!this.stopped) {
+			return TPromise.as([]);
+		}
+		if (!this.callStackQueried) {
```

Comment:
I believe the `callStackQueried` flag is superflous and we can remove it. What I propose is to just change line 126 to `if (!this.promisedCallStack)`. This way if we already queried a call to the adapter the original call will be returned and no additional calls will be made. Same as the `ExpressionsContainer` does with `getChildren`